src/typescript/meadow-project-manager.ts

```diff
@@ -382,4 +384,53 @@ export class MeadowProjectManager {
 
 		return false;
 	}
+
+	static launchConfiguration = vscode.workspace.getConfiguration('launch');
+	static savedConfigurations: any;
+
+	public static async refreshDeviceList(): Promise<void> {
+		var util = new MeadowUtil();
+		var meadowDevices : DeviceData[] = [];
+
+		await vscode.window.withProgress({
+			location: vscode.ProgressLocation.Notification,
+			cancellable: false,
+			title: 'Refreshing Device List'
+		}, async (progress) => {
+			// Load devices
+			progress.report({  increment: 0 });
+			meadowDevices = await util.GetDevices();
+			progress.report({ increment: 100 });
+		});
+
+		if (this.savedConfigurations === undefined){
+			this.savedConfigurations = this.launchConfiguration['configurations'];
+		}
+
+		let configurations = this.savedConfigurations;
```

Comment:
This doesn't work as intended and my typescript-fu or google-fu can't find the right answers, so that this isn't a reference assignment. `configurations` needs to be a local copy that doesn't modify `this.savedConfigurations` so that we can reset that file on shutdown.

Reply:
I've found a way to clone the `this.savedConfigurations` object in a recent commit.

Reply:
Glad you found that serialize/parse option. I was going to suggest the spread operator, but that would only avoid one level reference passing. Any sub-objects within would still be by reference.

Not that you want to use it because of that issue, but here's what that operator would look like.

```javascript
let configurations = { ...this.savedConfigurations };
```

That falls apart for sub-objects that are part of `this.savedConfigurations`, though. For a simplified example, this shows the issue I ran into.

```javascript
let x = { a: "a", b: 78, c: { ca: true, cb: "hello" } };
let y = { ...x };

x.b = 5; // doesn't impact `y` :)
x.c.ca = false; // does impact `y.c` :(

console.log(y);
//y: {
//  "a": "a",
//  "b": 78, // this was by value
//  "c": {
//    "ca": false, // this field's parent object was by reference, allowing upstream changes to affect it
//    "cb": "hello"
//  }
//}
```

Reply:
@patridge thanks for the explanation. Very useful!